Sources/MongoSwift/ReadPreference.swift

```diff
@@ -193,104 +186,73 @@ public struct ReadPreference {
-        self.readPref = mongoc_read_prefs_copy(pointer)
+        self.mode = mode
+        self.tagSets = nil
+        self.maxStalenessSeconds = nil
     }
 
-    fileprivate convenience init(mode: ReadPreference.Mode, tagSets: [Document]?, maxStalenessSeconds: Int?) throws {
-        self.init(mode)
-
-        if let tagSets = tagSets, !tagSets.isEmpty {
-            let tags = Document(tagSets.map { .document($0) })
-            tags.withBSONPointer { tagsPtr in
-                mongoc_read_prefs_set_tags(self.readPref, tagsPtr)
-            }
-        }
-
-        if let maxStalenessSeconds = maxStalenessSeconds {
-            guard maxStalenessSeconds >= MONGOC_SMALLEST_MAX_STALENESS_SECONDS else {
+    private init(_ mode: Mode, tagSets: [Document]?, maxStalenessSeconds: Int?) throws {
+        if let maxStaleness = maxStalenessSeconds {
+            guard maxStaleness >= MONGOC_SMALLEST_MAX_STALENESS_SECONDS else {
                 throw InvalidArgumentError(
                     message: "Expected maxStalenessSeconds to be >= " +
-                        " \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStalenessSeconds) given"
+                        " \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStaleness) given"
                 )
             }
-            mongoc_read_prefs_set_max_staleness_seconds(self.readPref, Int64(maxStalenessSeconds))
         }
+        self.mode = mode
+        self.tagSets = tagSets
+        self.maxStalenessSeconds = maxStalenessSeconds
     }
 
-    fileprivate var mode: ReadPreference.Mode {
-        ReadPreference.Mode(mongocMode: mongoc_read_prefs_get_mode(self.readPref))
-    }
+    /// Initializes a new `ReadPreference` by copying a `mongoc_read_prefs_t`. Does not free the original.
+    internal init(copying pointer: OpaquePointer) {
+        self.mode = Mode(mongocMode: mongoc_read_prefs_get_mode(pointer))
 
-    fileprivate var tagSets: [Document]? {
-        guard let bson = mongoc_read_prefs_get_tags(self.readPref) else {
+        guard let tagsPointer = mongoc_read_prefs_get_tags(pointer) else {
             fatalError("Failed to retrieve read preference tags")
         }
         // we have to copy because libmongoc owns the pointer.
-        let wrapped = Document(copying: bson)
+        let wrappedTags = Document(copying: tagsPointer)
+        if !wrappedTags.isEmpty {
+            // swiftlint:disable:next force_unwrapping
+            self.tagSets = wrappedTags.values.map { $0.documentValue! } // libmongoc will always return array of docs
+        }
 
-        guard !wrapped.isEmpty else {
-            return nil
+        let maxStalenessValue = mongoc_read_prefs_get_max_staleness_seconds(pointer)
+        if maxStalenessValue != MONGOC_NO_MAX_STALENESS {
+            self.maxStalenessSeconds = Int(exactly: maxStalenessValue)
         }
+    }
 
+    internal func withMongocReadPreference<T>(body: (OpaquePointer) throws -> T) rethrows -> T {
```

Comment:
probably worth documenting that the pointer is only valid within the closure and will need its contents copied if preserved.

Reply:
done